pkg/server/server.go

```diff
@@ -91,10 +94,37 @@ func NewServer(ctx context.Context, cfg *config.Config, logger *zap.Logger) (srv
 		api.Register(engine.Group("/api"), ready, cfg.API, logger.Named("api"), srv.NamespaceManager, srv.ConfigManager)
 
 		etcd_cfg := embed.NewConfig()
-		etcd_cfg.LCUrls = cfg.LCUrls
-		etcd_cfg.ACUrls = cfg.ACUrls
-		etcd_cfg.LPUrls = cfg.LPUrls
-		etcd_cfg.APUrls = cfg.APUrls
+
+		apiAddrStr := cfg.API.Addr
+		if !strings.HasPrefix(apiAddrStr, "http://") {
+			apiAddrStr = fmt.Sprintf("http://%s", apiAddrStr)
+		}
+		apiAddr, uerr := url.Parse(apiAddrStr)
+		if uerr != nil {
+			err = errors.WithStack(uerr)
+			return
+		}
+		etcd_cfg.LCUrls = []url.URL{*apiAddr}
+		apiAddrAdvertise := *apiAddr
+		apiAddrAdvertise.Host = fmt.Sprintf("%s:%s", pubAddr, apiAddrAdvertise.Port())
+		etcd_cfg.ACUrls = []url.URL{apiAddrAdvertise}
+
+		peerPort := cfg.Advance.PeerPort
+		if peerPort == "" {
+			peerPortNum, uerr := strconv.Atoi(apiAddr.Port())
+			if uerr != nil {
+				err = errors.WithStack(uerr)
+				return
+			}
+			peerPort = strconv.Itoa(peerPortNum + 1)
+		}
+		peerAddr := *apiAddr
+		peerAddr.Host = fmt.Sprintf("%s:%s", peerAddr.Hostname(), peerPort)
+		etcd_cfg.LPUrls = []url.URL{peerAddr}
+		peerAddrAdvertise := *apiAddr
+		peerAddrAdvertise.Host = fmt.Sprintf("%s:%s", pubAddr, peerPort)
+		etcd_cfg.APUrls = []url.URL{peerAddrAdvertise}
+
```

Comment:
Extract the etcd part to a new function.

Reply:
Done.